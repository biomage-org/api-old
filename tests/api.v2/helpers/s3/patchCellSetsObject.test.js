// @ts-nocheck
const patchCellSetsObject = require('../../../../src/api.v2/helpers/s3/patchCellSetsObject');
const getObject = require('../../../../src/api.v2/helpers/s3/getObject');
const putObject = require('../../../../src/api.v2/helpers/s3/putObject');

jest.mock('../../../../src/api.v2/helpers/s3/getObject');
jest.mock('../../../../src/api.v2/helpers/s3/putObject');

const mockExperimentId = 'mock-experiment-id';

const mockCellSets = JSON.stringify({
  cellSets: [
    {
      key: 'louvain',
      name: 'louvain clusters',
      rootNode: true,
      type: 'cellSets',
      children: [
        {
          key: 'louvain-0',
          name: 'Cluster 0',
          rootNode: false,
          type: 'cellSets',
          color: '#77aadd',
          cellIds: [0, 1, 2, 3],
        },
      ],
    },
  ],
});

const mockPatch = [
  {
    $match: {
      query: '$[?(@.key == "louvain")]',
      value: {
        children: [
          {
            $insert:
              {
                index: '-',
                value:
                  {
                    key: 'new-cluster-1',
                    name: 'New Cluster 1',
                    rootNode: false,
                    color: '#3957ff',
                    type: 'cellSets',
                    cellIds: [4, 5, 6],
                  },
              },
          },
        ],
      },
    },
  },
];

getObject.mockReturnValue(mockCellSets);

describe('patchCellSetsObject', () => {
  it('Works correctly', async () => {
    const result = await patchCellSetsObject(mockExperimentId, mockPatch);

    // Put a modified object
    const putParams = putObject.mock.calls[0][0];

    expect(putParams).toMatchSnapshot();

    // Does not return anything on success
    expect(result).toBeUndefined();
  });

  it('Throws an error if the JSON merger result is not correct', async () => {
    // Should fail validation because cellIds is not an array
    const malformedPatch = [
      {
        $match: {
          query: '$[?(@.key == "louvain")]',
          value: {
            children: [
              {
                $insert:
              {
                index: '-',
                value:
                  {
                    key: 'singular-cluster',
                    name: 'Singular cluster',
                    rootNode: false,
                    color: '#3957ff',
                    type: 'cellSets',
                    cellIds: 1,
                  },
              },
              },
            ],
          },
        },
      },
    ];

    await expect(patchCellSetsObject(mockExperimentId, malformedPatch)).rejects.toThrow();
  });
});
