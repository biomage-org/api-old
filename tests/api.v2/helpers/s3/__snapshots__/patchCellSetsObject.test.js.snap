// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`patchCellSetsObject Works correctly 1`] = `
Object {
  "Body": "{\\"cellSets\\":[{\\"key\\":\\"louvain\\",\\"name\\":\\"louvain clusters\\",\\"rootNode\\":true,\\"type\\":\\"cellSets\\",\\"children\\":[{\\"key\\":\\"louvain-0\\",\\"name\\":\\"Cluster 0\\",\\"rootNode\\":false,\\"type\\":\\"cellSets\\",\\"color\\":\\"#77aadd\\",\\"cellIds\\":[0,1,2,3]},{\\"key\\":\\"new-cluster-1\\",\\"name\\":\\"New Cluster 1\\",\\"rootNode\\":false,\\"color\\":\\"#3957ff\\",\\"type\\":\\"cellSets\\",\\"cellIds\\":[4,5,6]}]}]}",
  "Bucket": "cell-sets-test-000000000000",
  "Key": "mock-experiment-id",
}
`;
